networkx/algorithms/dominating.py

```diff
@@ -93,3 +102,154 @@ def is_dominating_set(G, nbunch):
     testset = {n for n in nbunch if n in G}
     nbrs = set(chain.from_iterable(G[n] for n in testset))
     return len(set(G) - testset - nbrs) == 0
+
+
+@not_implemented_for("directed")
+@nx._dispatchable
+def connected_dominating_set(G):
+    """Returns a connected dominating set.
+
+    A *dominating set* for a graph *G* with node set *V* is a subset *D* of
+    *V* such that every node not in *D* is adjacent to at least one
+    member of *D* [1]_. A *connected dominating set* is a dominating
+    set *C* that induces a connected subgraph of *G* [2]_.
+
+    Parameters
+    ----------
+    G : NewtorkX graph
+        Undirected connected graph.
+
+    Returns
+    -------
+    connected_dominating_set : set
+        A dominating set of nodes which induces a connected subgraph of G.
+
+    Examples
+    ________
+    >>> G = nx.Graph(
+    ...     [
+    ...         (1, 2),
+    ...         (1, 3),
+    ...         (1, 4),
+    ...         (1, 5),
+    ...         (1, 6),
+    ...         (2, 7),
+    ...         (3, 8),
+    ...         (4, 9),
+    ...         (5, 10),
+    ...         (6, 11),
+    ...         (7, 12),
+    ...         (8, 12),
+    ...         (9, 12),
+    ...         (10, 12),
+    ...         (11, 12),
+    ...     ]
+    ... )
+    >>> nx.connected_dominating_set(G)
+    {1, 2, 3, 4, 5, 6, 7}
+
+    Raises
+    ------
+    NetworkXNotImplemented
+        If G is directed.
+
+    NetworkXError
+        If G is disconnected.
+
+    Notes
+    -----
+    This function implements Algorithm I in its basic version as described
+    in [3]_. The idea behind the algorithm is the following: grow a tree *T*,
+    starting from the node of maximum degree. Nonleaf nodes in *T* are our
+    connected dominating set (CDS), leaf nodes in *T* are marked as seen and all
+    other nodes are marked as unseen. At each step we pick a seen (leaf) node
+    with maximal number of unseen neighbors, add it to the CDS and mark all its
+    neighbors as seen. The algorithm terminates when there are no more unseen
+    nodes.
```

Comment:
I think this paragraph is not quite right in how neighbors are selected. What do you think of the following?
```suggestion
    This function implements Algorithm I in its basic version as described
    in [3]_. The idea behind the algorithm is the following: grow a spanning tree *T*
    starting from a node with maximum degree and add neighbors based on
    maximal connectivity as follows. Track the degree of each node in the subgraph
    induced by nodes not yet in *T*. At each step, add the node with maximal
    subgraph degree of those already "seen", that is, they are a neighbor of
    some node in *T*. While adding the node to *T*, decrement the degree of
    its neighbors and add them to the heap to be considered as the next node
    to add. The process ends when all "seen" neighbors have been added. 
    The connected dominating set (CDS) is all non-leaf nodes in *T*.
```

Reply:
A couple of notes:
 1. Our "goal" is to move as many nodes as possible from "unseen" to "seen" in each iteration (note that in each iteration only one node can move from "seen" to the CDS). That's why we keep track of the number of "unseen" neighbors every node has. 
 2. A node in the CDS has no neighbors in "unseen", since we moved those neighbors to "seen" when we moved the node from "seen" to the CDS.
 3. T consists of all nodes in the CDS and the nodes in "seen", so the sentence "Track the degree...." in lines 164-165 is inaccurate, since a node in "seen" may have a neighbor in "seen" but that connection isn't reflected in the tracking of the "unseen-degree" (that's why I named it so). I don't think there's any added value in referring to the "subgraph induced by nodes not yet in T".
 
I think that what I wrote reflects more accurately what's going on in the algorithm, unless I'm not understanding correctly what the code I wrote does. Although it would make it clearer to mention the degree-tracking mechanism.
So I would suggest the following:
`This function implements Algorithm I in its basic version as described
in [3]_. The idea behind the algorithm is the following: grow a tree *T*,
starting from a node with maximum degree. Throughout the growing process,
nonleaf nodes in *T* are our connected dominating set (CDS), leaf nodes in *T* 
are marked as "seen" and nodes in G that are not yet in *T* are marked as "unseen".
We maintain a max-heap of all "seen" nodes, and track the number of "unseen" neighbors
for each node. At each step we pop the heap top--a "seen" (leaf) node with maximal
number of "unseen" neighbors, add it to the CDS and mark all its "unseen" neighbors as "seen".
For each one of the newly created "seen" nodes, we also decrement the number of "unseen" neighbors
for all its neighbors.
The algorithm terminates when there are no more unseen nodes.`

Sorry for the formatting, couldn't figure out how to have it look like yours.

Thanks.

Reply:
I thought the three categories were CDS nodes, seen nodes, unseen nodes. And I thought they were **distinct** sets.  That is, `seen` includes neighbors of CDS nodes that are not in CDS. `unseen` includes nodes that are not neighbors of nodes in CDS. CDS includes nodes in the tree.

I think you are saying that nodes in `seen` are also in the tree. Don't they only get added to the tree when they are popped from the queue?

Reply:
> I think you are saying that nodes in seen are also in the tree.

Exactly.

> Don't they only get added to the tree when they are popped from the queue?

 They are added to the tree as soon as they are pushed into "seen" and are added to the CDS when they are popped. The tree isn't explicit in the algorithm. The CDS and "seen" serve the purpose of separating leaf and nonleaf nodes of the tree. In each step we want to add as many nodes as possible to the tree and this can only be done by considering the leaf nodes.

Hope that clears things up.

Reply:
Thanks for that!  
I knew we weren't actually creating the tere, but I didn't have the right model in my head for how the sets related to the tree. "seen" nodes are the current leaves in the tree. "unseen" are not in the tree, and CDS are the nodes that dominate the tree.  That way we don't actually construct the tree and then remove the leaves. They are already split apart.

Your suggestion from a couple comments up looks good to me. 

I'll try to make it a "suggestion" by selecting the lines for the suggestion and adding a comment that starts with "```suggestion" on its own line and ends with "```". [see this stackoverflow or search for github suggestion](https://stackoverflow.com/questions/54239733/how-can-i-manually-add-suggestions-in-code-reviews-on-github)

```suggestion
    This function implements Algorithm I in its basic version as described
    in [3]_. The idea behind the algorithm is the following: grow a tree *T*,
    starting from a node with maximum degree. Throughout the growing process,
    nonleaf nodes in *T* are our connected dominating set (CDS), leaf nodes in *T* 
    are marked as "seen" and nodes in G that are not yet in *T* are marked as "unseen".
    We maintain a max-heap of all "seen" nodes, and track the number of "unseen"
    neighbors for each node. At each step we pop the heap top -- a "seen" (leaf) node
    with maximal number of "unseen" neighbors, add it to the CDS and mark all its
    "unseen" neighbors as "seen". For each one of the newly created "seen" nodes,
    we also decrement the number of "unseen" neighbors for all its neighbors.

    The algorithm terminates when there are no more "unseen" nodes.
```

Reply:
Glad we got that cleared up. 
So should I change the description according to the previous comment and remove the `pytest.param` from the test and push?

Reply:
Yes -- I think that will resolve all existing conversations. :)

Reply:
Great. Pushed.
I sincerely appreciate your patience and guidance.
Thanks Dan and Alejandro.